### PR TITLE
Implement GET API for System Feature Upgrades

### PR DESCRIPTION
**Pull request by williamrandolph**
Here is an implementation for the GET API for system feature upgrades. It is pretty simple. When called, we iterate over all features, resolve all indices, and fetch the Version for each index from cluster state index metadata. We then determine the earliest index creation version for each feature to determine whether the features need to be upgraded or not.

I've also added a simple BWC test for rolling upgrades to make sure that we are in fact returning the correct index creation versions.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java [hunk 1]
```diff
@@ -60,16 +64,50 @@ public TransportGetFeatureUpgradeStatusAction(
...
+            .map(entry -> getFeatureUpgradeStatus(state, entry))
+            .collect(Collectors.toList());
+
+        boolean isUpgradeNeeded = features.stream()
+                .map(GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus::getMinimumIndexVersion)
+                .min(Version::compareTo)
+                .orElse(Version.CURRENT)
+                .before(Version.V_7_0_0);
+
+        listener.onResponse(new GetFeatureUpgradeStatusResponse(features, isUpgradeNeeded ? "UPGRADE_NEEDED" : "NO_UPGRADE_NEEDED"));
```
Comment (gwbrown) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java:78]: Nit: Can you move `"UPGRADE_NEEDED"` and `"NO_UPGRADE_NEEDED"` into a constant?

File: qa/system-indices/src/javaRestTest/java/org/elasticsearch/system/indices/FeatureUpgradeApiIT.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        assertThat(testFeature.size(), equalTo(4));
+        assertThat(testFeature.get("minimum_index_version"), equalTo(Version.CURRENT.toString()));
+        assertThat(testFeature.get("upgrade_status"), equalTo("NO_UPGRADE_NEEDED"));
+        assertThat(testFeature.get("indices"), instanceOf(List.class));
+
+        assertThat((List<Object>) testFeature.get("indices"), hasSize(1));
+    }
+
+    @After
+    public void resetFeatures() throws Exception {
```
Comment (gwbrown) [inline qa/system-indices/src/javaRestTest/java/org/elasticsearch/system/indices/FeatureUpgradeApiIT.java:67]: Nit: Could you move this to before any of the test methods? It's nice to get a handle on the suite environment up front.

File: server/src/internalClusterTest/java/org/elasticsearch/migration/FeatureUpgradeIT.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.elasticsearch.action.admin.cluster.migration.GetFeatureUpgradeStatusResponse;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class FeatureUpgradeIT extends ESIntegTestCase {
+
+    public void testGetFeatureUpgradeStatus() throws Exception {
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/migration/FeatureUpgradeIT.java:23]: What does this test buy us over the ones in `FeatureUpgradeApiIT`?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java [hunk 4]
```diff
@@ -60,16 +64,50 @@ public TransportGetFeatureUpgradeStatusAction(
...
+    }
+
+    private GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus getFeatureUpgradeStatus(
```
Comment (gwbrown) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java:81]: Could you add a couple unit tests for this method? I like the integration tests in this PR, but none of the test the path that ends up returning `UPGRADE_NEEDED` - doesn't have to be terribly involved, just a quick check is good enough I think.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+public class FeatureUpgradeIT extends AbstractRollingTestCase {
+
+    @SuppressWarnings("unchecked")
+    public void testGetFeatureUpgradeStatus() throws Exception {
+
+        final String systemIndexWarning = "this request accesses system indices: [.tasks], but in a future major version, direct " +
+            "access to system indices will be prevented by default";
+        if (CLUSTER_TYPE == ClusterType.OLD) {
+            // setup - put something in the watcher index
```
Comment (gwbrown) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java:30]: This test doesn't do anything to `.watches` - left over from a copy/paste?

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            getTasksIndex.setOptions(expectVersionSpecificWarnings(v -> {
+                v.current(systemIndexWarning);
+                v.compatible(systemIndexWarning);
+            }));
+            getTasksIndex.addParameter("allow_no_indices", "false");
+
+            getTasksIndex.setOptions(expectVersionSpecificWarnings(v -> {
+                v.current(systemIndexWarning);
+                v.compatible(systemIndexWarning);
+            }));
```
Comment (gwbrown) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java:73]: Did you mean to set the options twice here?

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                    entityAsMap(client().performRequest(clusterStateRequest)));
+
+                List<Map<String, Object>> features = view.get("features");
+                Map<String, Object> feature = features.stream()
+                    .filter(e -> "tasks".equals(e.get("feature_name")))
+                    .findFirst()
+                    .orElse(Collections.emptyMap());
+
+                assertThat(feature.size(), equalTo(4));
+                assertThat(feature.get("minimum_index_version"), equalTo(UPGRADE_FROM_VERSION.toString()));
```
Comment (gwbrown) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java:96]: I suspect we could do some logic on the old and new versions and check the `(NO_)UPGRADE_NEEDED` bit as well, if the old version is a different major than the new version.

**Review by AthenaEryma — COMMENTED** [review]
Thanks, this is looking good, especially the production code! Other than a super minor nit, my only comments were on the tests.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java (hunk 1 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java:79]: Good call. I should have done that in the API stubs PR.

Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java]: I used an enum, but let me know if you think a pair of string constants would be better.

**Comment by sabarasaba** [comment]
Great work @williamrandolph! I have a few questions regarding `upgrade_status` that I couldn't figure out from looking at the code:

1.  If upgrading and index fails, will the upgrade_status of the feature change to something like `ERROR`? If so, is there going to be a way to have more details about why it failed?
2. What are all the possible values upgrade_status can have? So far I've noted `'UPGRADE_NEEDED' | 'NO_UPGRADE_NEEDED' | 'IN_PROGRESS' | 'ERROR'` but not sure about the last two

**Review by AthenaEryma — APPROVED** [review]
LGTM!

**Pull request merged**